### ISSUE TITLE
Explore E programming language and capability transport

### ISSUE THREAD

**Issue by sergefdrv** [replica-io/replica-io#67]
Explore [E distributed programming language](https://en.wikipedia.org/wiki/E_(programming_language)) and the [capability transport protocol](http://www.erights.org/elib/distrib/captp/)

**Comment by sergefdrv** [comment]
[E](http://erights.org/) is a secure distributed programming language based on the [object-capability model](https://en.wikipedia.org/wiki/Object-capability_model). It is _purely object-oriented_ with Java-like syntax and dynamic typing. Distributed programming in E is relies on _semi-transparent networking_. It has implementations based on Java and Common Lisp.

Being a purely object-oriented language, E follows the principle "everything is an object". So the central concept in E's programming model is the notion of an object that has internal local state and exhibits some behavior. An object can be either a primitive (e.g. literal number) or an instance. An non-primitive instance object is a combination of code and state, where state is a mutable collection of references to objects. Objects form a dynamic reference graph _interacting by sending messages on references_. Messages can carry references to objects as arguments, thus causing changes in the connectivity of the reference graph.

Object references in E are _unforgeable_ and act at the same time as communication channels and as _transferable capabilities_ to access designated objects. Unforgeability of references is achieved with unguessable numbers ("Swiss numbers") and their hashes. This allows building secure systems following the [principle of least authority (PoLA)](https://en.wikipedia.org/wiki/Principle_of_least_privilege). E also provides a primitive for securely establishing _immediate equality_ of remote object references; using this primitive mechanisms for establishing _eventual reference equality_ can be built, e.g. the trust combining mechanism `E.join`. References (capabilities) can be obtained by objects only through the following means: _initial conditions_ (supplied from outside), _parenthood_ (by the parent object when instantiating a child), _endowment_ (by capturing references from the parent by the child when being instantiated), _introduction_ (by explicit communication between objects).

Syntactically, objects in E are defined as lambda abstractions ([lexical closures](https://en.wikipedia.org/wiki/Closure_(computer_programming))) with message dispatch hooks (methods) producing local side effects in response to reception of messages by the object. The references captured by the object's closure constitute the object's local state. Mutable state is represented by variables that are realized as names bound to a special kind of object called _slots_.

Objects interact by sending messages on references. Each message is a combination of a verb  and arguments. The verb indicates the method to be executed with the arguments by the target object upon reception of the message. E provides two message sending primitives: immediate and eventual send. The _immediate send_ represents the usual method call that is immediately executed and blocks the sender until the recipient returns the control and the result back; immediate calls are only allowed on local objects (near references). The _eventual send_ is a non-blocking primitive (the _eventually operator_ `<-`) that arranges for the message to be eventually delivered to the target object and returns back immediately with a _promise_ for the outcome; eventual sends can be done on any reference.

_Promises_ are a form of _delayed reference_ to the result of an eventual send. Once the outcome of the eventual send is known, the corresponding promise becomes _resolved_. Eventually the promise becomes a reference designating the outcome, called the _resolution_. To avoid cascading round trips, E supports _promise pipelining_: promises can be immediately used as an arguments or even as a targets for eventual sends. Because messages can be eventually sent to a promise even before it is resolved, they are buffered in FIFO order within the promise until it is resolved and eventually forwarded to the resolution. An example of promise pipelining is depicted on the image below.

![Image](https://github.com/user-attachments/assets/dd90aa31-f36e-4988-9d90-cfc3f88e78c3)
_(Image source: http://erights.org/talks/thesis/markm-thesis.pdf)_

There are different states that a reference can be in. First of all, a reference can be either unresolved or resolved. An _unresolved_ reference is a _promise_, which can be either local or remote, depending on whether it designates an outcome of an eventual send on a local or remote object. An unresolved promise eventually becomes _resolved_ to either near or far reference. Local promises are also referred to as _vows_; promises that can be either local or remote are referred to as _receivers_. A _near reference_ designates a local object, and can be used for both immediate and eventual sends, whereas a _far_ reference designates a remote object and can only be used for eventual sends. A _broken_ reference designates a problem with fulfilling the promise. A promise can resolve into a broken reference in case of an exception thrown by the target object's method. A far reference can become broken in case of a partial failure, e.g. network partition. A broken reference remains broken forever. The reference states and transitions described above are depicted on the diagram below.

![Image](https://github.com/user-attachments/assets/4f2bc128-78f7-4384-8e86-0cf761aa4610)
_(Image source: http://erights.org/talks/thesis/markm-thesis.pdf)_

The reference described above are referred to as _live references_, and they guarantee _ordered_ delivery of messages eventually send on them. There is another kind or reference called _sturdy_ reference that cannot be used for sending messages. However, unlike live reference, a sturdy reference is not ephemeral since it holds a URI to the designated object and can be used to create new live references to the object whenever needed.

There is a dedicated syntax, the when-catch expression, that can be used to attach a continuation to a promise. The body of the when-catch expression is executed once the promise becomes resolved. The main body is executed is executed when the promise is fulfilled, i.e. resolves to a near or far reference. If the promise resolves to a broken reference than the catch clause is executed instead. When-catch can have a finally clause that executes after either the main body or the catch clause. The when-catch expression can be multi-way, i.e. can have multiple promises as arguments and wait until all of them become resolved. When-catch expressions are asynchronous, i.e. the execution of the code containing a when-catch continues without blocking. The when-catch expression itself returns a promise that is resolved after its body is executed.

Concurrency in E is based on _communicating event loops_. The process-like unit of execution is called _vat_. Each vat executes concurrently with all other vats, but there is only a single thread of execution within a vat executing its _event loop_. Each vat has a _heap of objects_, a _call-return stack_, and a _queue of pending deliveries_. The vat structure is depicted on the image below.

![Image](https://github.com/user-attachments/assets/64f4d376-81e3-414c-b2c6-999bb1b9ed07)
_(Image source: http://erights.org/talks/thesis/markm-thesis.pdf)_

 An immediate call pushes a new stack frame, whereas an eventual send arranges for a new pending delivery, which is a pair of a message and the target object, to be enqueued onto the pending delivery queue of the target object's vat. Thus there are _both a normal execution stack and a queue_ containing the pending deliveries. On each iteration, the vat’s event loop dequeues the next pending delivery from the queue and invokes the target with the message. This top-level invocation, called a _turn_, is _atomic_ because it executes to completion before the next event loop iteration begins, i.e. _vat's turns are temporally isolated_. The event-loop concurrency model employed in E eliminates classical deadlocks, though dataflow locks caused by circular data dependencies between promises remain possible.
 
E guarantees a _partial ordering_ of messages, called _E-order_, which is stronger than FIFO but weaker than casual order. First of all, the messages eventually sent on the same reference are all delivered in the order sent unless and until the reference becomes broken; after that, no more message is delivered. A delivery of a message is guaranteed to happen only after all messages sent earlier on the same reference have been already delivered. However, when a reference is eventually sent in message, the reference being communicated is considered forked. So the target object receives a fork of the reference created at the time when the eventual send was performed. When a fork occurs between two sending events, the fork of the reference designates the target after the first message, i.e. that has already received the message sent to it before the forking.  The tree of messages connected by the reference topology according to these precedence rules forms a partial order. An example of such ordering is depicted on the image below.

![Image](https://github.com/user-attachments/assets/9962bbbd-9184-465f-86a1-0b25f894687f)
_(Image source: http://erights.org/talks/thesis/markm-thesis.pdf)_

E is a dynamically typed language with relatively simple type system. Variables, promises, and return values can be optionally annotated with _guards_. Guards act as a kind of _soft type-checking_ that is performed _at runtime_. There is a number of built-in guards, sub-ranges and more complex expressions can also act as guards, and one can also define custom guards (interfaces). Transparent immutable objects (e.g. immutable built-in data types) are always passed by copy and are guaranteed to resolve to local objects. There is a more general notion of an object _passed by construction_ which, when sent over a comm system, is sent as some serialized representation. Mutable objects, on the other hand, are always _passed by proxy_ (as a promise that resolves into a far reference). E supports a form of inheritance by allowing an object to delegate some methods to another object, a super-object, and optionally passing a self-reference to the super-object's constructor.

Inter-process capability-secure distributed programming in E is provided by [ELib](http://erights.org/elib/), a Java library. The _cryptographic capability protocol_ implemented in ELib allows mutually-suspicious processes to cooperate safely. It also implements efficient event-loop concurrency and promise pipelining. Vats can be made persistent with periodic checkpoints, saving the vat's persistent state to non-volatile storage. A persistent vat can be revived from the latest checkpoint after a crash. A crash partitions a vat from all others, and old vat-crossing references will be revived as broken, so only offline capabilities (sturdy references) in either direction enable it to become reconnected.

The Capability Transport Protocol, [CapTP](erights.org/elib/distrib/captp/), allows object references to span across vats so that they follow the same dynamic capability logic as they do within a vat. In order to achieve this, CapTP maintains a set of _tables_ that keep track of vat-crossing references as positions in those tables. There are two dual kinds of tables maintained between each pair of vats in each direction: question and answer tables keep track of positions allocated by the party holding the outgoing (tail) end of the reference, whereas export and import tables keep track of positions allocated by the party holding the incoming (arrow) end of the reference. In order to ensure the partial order guarantees in case of promise pipelining crossing boundaries of more than two vats, CapTP implements a three-vat protocol is layered on the two-vat protocol. Apart from the questing/answer and export/import tables, each vat maintains a _nonce locator_ for each inter-vat CapTP connection, wich is established at most once for every pair of vats. The nonce locator is used to keep track of forked references as they are passed between vats. The CapTP protocol is also responsible for distributed garbage collection of inter-vat references that are no longer needed.
